### PR TITLE
Finish login

### PR DESCRIPTION
**Pull request by hellovietnam93**

### CODE REVIEW

**Comment by minhld1709** [comment]
OK

**Comment by VuongLong** [comment]
OK

**Comment by ttuan** [comment]
OK

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,27 @@
 class ApplicationController < ActionController::Base
-  # Prevent CSRF attacks by raising an exception.
-  # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  before_action :authenticate_user!
+
+  protected
+  def verify_admin
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: use `cancancan` instead of this

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
+    unless current_user.admin?
+      flash[:danger] = t "flash.access_denied"
+      redirect_to root_path
+    end
+  end
+
+  def verify_supervisor
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: use `cancancan` instead of this

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,5 +1,27 @@
...
+    unless current_user.supervisor?
+      flash[:danger] = t "flash.access_denied"
+      redirect_to root_path
+    end
+  end
+
+  def logged_in_user
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: use `cancancan` instead of this

File: app/models/user.rb [hunk 4]
```diff
@@ -6,4 +6,14 @@ class User < ActiveRecord::Base
   has_many :courses, through: :user_courses
   has_many :subjects, through: :user_subjects
   has_many :tasks, through: :user_tasks
+
+  after_initialize :set_default_role, if: :new_record?
```
Comment (dieunb) [inline app/models/user.rb]: override `initialize` function instead of this

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+Devise.setup do |config|
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (dieunb) [inline config/initializers/devise.rb]: Don't remove all config of devise, should restore it

**Pull request merged**